File: packages/scanner/src/scoring.ts
/**
 * Scoring utilities for risk aggregation. Part of the Aegis token-scanning skill.
 *
 * Severity weighting mirrors the firewall's, so a trader reads one consistent
 * scale across both gates.
 */

import type { CheckResult, Severity } from "./types.js";

export const SEVERITY_WEIGHT: Record<Severity, number> = {
  ok: 0,
  info: 1,
  warn: 3,
  danger: 10,
};

export function riskLevel(score: number, caution = 25, nogo = 60): "low" | "medium" | "high" {
  if (score >= nogo) return "high";
  if (score >= caution) return "medium";
  return "low";
}

export function dangerCount(results: CheckResult[]): number {
  return results.filter((r) => r.severity === "danger").length;
}

export function warnCount(results: CheckResult[]): number {
  return results.filter((r) => r.severity === "warn").length;
}

export function passingCount(results: CheckResult[]): number {
  return results.filter((r) => r.severity === "ok").length;
}

/** Group cumulative score by check family (the prefix before the first "-"). */
export function scoreBreakdown(results: CheckResult[]): Record<string, number> {
  const out: Record<string, number> = {};
  for (const r of results) {
    const area = r.check.includes("-") ? r.check.split("-")[0]! : r.check;
    out[area] = (out[area] ?? 0) + r.score;
  }
  return out;
}

File: packages/scanner-py/src/hoodtrade/scoring.py
"""Scoring utilities for risk aggregation.

Provides helpers for weighted scoring, score normalization, and risk
categorization beyond the simple sum used in the engine.
"""

from __future__ import annotations

from .models import CheckResult, Severity

SEVERITY_WEIGHT = {
    Severity.OK: 0,
    Severity.INFO: 1,
    Severity.WARN: 3,
    Severity.DANGER: 10,
}


def weighted_score(results: list[CheckResult]) -> float:
    if not results:
        return 0.0
    total = sum(r.score * SEVERITY_WEIGHT.get(r.severity, 1) for r in results)
    return total / len(results)


def normalize_score(score: int, max_score: int = 200) -> float:
    return min(score / max_score, 1.0) if max_score > 0 else 0.0


def risk_level(score: int, caution: int = 25, nogo: int = 60) -> str:
    if score >= nogo:
        return "high"
    if score >= caution:
        return "medium"
    return "low"


def danger_count(results: list[CheckResult]) -> int:
    return sum(1 for r in results if r.severity is Severity.DANGER)


def warn_count(results: list[CheckResult]) -> int:
    return sum(1 for r in results if r.severity is Severity.WARN)


def passing_count(results: list[CheckResult]) -> int:
    return sum(1 for r in results if r.severity is Severity.OK)


def score_breakdown(results: list[CheckResult]) -> dict[str, int]:
    breakdown: dict[str, int] = {}
    for r in results:
        area = r.check.split("-")[0] if "-" in r.check else r.check
        breakdown[area] = breakdown.get(area, 0) + r.score
    return breakdown
